### PR TITLE
rustdoc: skip `// some variants omitted` if enum is `#[non_exhaustive]`

### LINKED ISSUE #108925: Consider omitting "some variants omitted" in enum documentation if enum is non_exhaustive
<table>
<tr><th>code</th><th>rustdoc</th></tr>
<tr><td>

```rust
pub enum Enum1 {
    A,
    B,
    #[doc(hidden)]
    C,
}
```
</td><td></td></tr>
<tr><td>

```rust
#[non_exhaustive]
pub enum Enum2 {
    A,
    B,
}
```
</td><td></td></tr>
<tr><td>

```rust
#[non_exhaustive]
pub enum Enum3 {
    A,
    B,
    #[doc(hidden)]
    C,
}
```
</td><td></td></tr>
</table>

Would it be reasonable for Rustdoc to produce exactly the same output for `Enum2` and `Enum3`? I think so. These two types have an identical public API.

Separately, would it be reasonable for Rustdoc to produce exactly the same output for `Enum1` and `Enum2`? Whether that means printing a synthetic `#[non_exhaustive]` attribute on `Enum1` instead of the `// some variants omitted` comment, or some other approach that unifies all three of these outputs.

### Meta

`rustc --version --verbose`:
```console
rustc 1.70.0-nightly (e3dfeeaa4 2023-03-07)
binary: rustc
commit-hash: e3dfeeaa45f117281b19773d67f3f253de65cee1
commit-date: 2023-03-07
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 15.0.7
```

### PR DESCRIPTION
**Pull request by Ezrashaw**
Fixes #108925 

Never touched rustdoc before so probably not the best code.

cc @dtolnay

### CODE REVIEW

**Comment by Ezrashaw** [comment]
@jsha I've added the lint and implemented your advice in the linked issue.

As per the linked issue, `Enum1` warns with the new lint. `Enum2` and `Enum3` have identical output, just the `#[non_exhaustive]` attribute.

@rustbot ready

File: src/doc/rustdoc/src/lints.md [hunk 1]
```diff
@@ -374,3 +374,36 @@ warning: this URL is not a hyperlink
 
 warning: 2 warnings emitted
 ```
+
+## `synthetic_non_exhaustive`
```
Comment (camelid) [inline src/doc/rustdoc/src/lints.md:378]: It's not clear what this lint does based on its name. Perhaps a name along the lines of `exhaustive_but_hidden` would be better?

File: src/doc/rustdoc/src/lints.md [hunk 2]
```diff
@@ -374,3 +374,36 @@ warning: this URL is not a hyperlink
...
+
+This lint is **warn-by-default**. It detects enums with variants that are stripped (`#[doc(hidden)]`) but not marked with the `#[non_exhaustive]` attribute.
```
Comment (camelid) [inline src/doc/rustdoc/src/lints.md:380]: "stripped" is internal rustdoc jargon that we don't usually use in docs.

```suggestion
This lint is **warn-by-default**. It detects enums with variants that are `#[doc(hidden)]` but not marked with the `#[non_exhaustive]` attribute.
```

**Review by camelid — COMMENTED** [review]
I think this lint may need a crater run to make sure the warning doesn't cause too much breakage. I'm also not familiar with the motivation for this change, but in some ways, I think the lint may be a better fit for clippy. The reason is I'm a bit concerned it could be triggered by valid code.

**Comment by Ezrashaw** [comment]
@camelid I'm not convinced by this change now, it breaks `libcore` which obviously has `-D warnings`. While it's impossible to merge this PR, I do think that `libcore` and co have bad code in this regard. For example, take this snippet from `libbacktrace`:
```rust
/// The styles of printing that we can print
#[derive(Copy, Clone, Eq, PartialEq)]
pub enum PrintFmt {
    /// Prints a terser backtrace which ideally only contains relevant information
    Short,
    /// Prints a backtrace that contains all possible information
    Full,
    #[doc(hidden)]
    __Nonexhaustive,
}
```
It's just wrong and the exact thing this lint is for.

Not sure where you want to go from here.

**Comment by camelid** [comment]
@Ezrashaw Thanks for looking into the breakage! I do think this lint could potentially be a good fit for clippy (though they'd have to be consulted of course).

I'd also like to hear what @jsha's thoughts are on this since he proposed the lint: Do you think moving it to clippy would make sense?

**Comment by jsha** [comment]
I have to admit I don't know the factors that would favor a rustdoc lint or a clippy lint. Above you mention that it could be triggered by valid code. Are rustdoc lints only for input that is definitely invalid?

**Comment by Ezrashaw** [comment]
@jsha Not sure if someone else was going to answer this; nobody else has so here goes: I think that possibly valid code should *never* be triggered by (warn-by-default) rustdoc lints, period. I think that the question here is really: is code valid if it contains `#[doc(hidden)]` variants whilst not being `#[non_exhaustive]`. I am of the opinion that such code is invalid, but `libcore` contains code which triggers the lint (+ `-D warnings` so we can't compile the standard library).

The bottom line is: there is nothing we can feasibly do while this lint exists to get it to pass CI. I'll remove the lint and maybe we can land the one-line change which is the other part of this PR..

**Review by notriddle — APPROVED** [review]

**Comment by Ezrashaw** [comment]
@notriddle Agreed, I've removed that part of the PR, now there is just the one liner fix which the linked issue was primarily concerned with.

**Comment by Ezrashaw** [comment]
@jsha Can we land this now?

**Comment by camelid** [comment]
I updated the title since I believe the content of this PR has changed.

**Comment by GuillaumeGomez** [comment]
Please add a test for it and then I'll approve it. Also, if not done already, I think adding this as a clippy lint would be nice. Please at least open an issue there so it's not forgotten.
